File: samples/java/clustered-chat/app/chat/ChatEventSerializer.java
package chat;

import akka.actor.ExtendedActorSystem;
import akka.serialization.SerializerWithStringManifest;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.val;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

public class ChatEventSerializer extends SerializerWithStringManifest {
  private final int identifier;
  private final ObjectMapper mapper = new ObjectMapper();

  public ChatEventSerializer(ExtendedActorSystem system) {
    identifier = system.settings().config().getInt("akka.actor.serialization-identifiers.\"" + getClass().getName() + "\"");
  }

  @Override
  public int identifier() {
    return identifier;
  }

  @Override
  public String manifest(Object o) {
    if (o instanceof ChatEvent.ChatMessage) {
      return "M";
    } else if (o instanceof ChatEvent.JoinRoom) {
      return "J";
    } else if (o instanceof ChatEvent.LeaveRoom) {
      return "L";
    } else {
      throw new RuntimeException("Don't know how to serialize " + o);
    }
  }

  @Override
  public byte[] toBinary(Object o) {
    val baos = new ByteArrayOutputStream();
    try {
      mapper.writeValue(baos, o);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return baos.toByteArray();
  }

  @Override
  public Object fromBinary(byte[] bytes, String manifest) {
    try {
      switch(manifest) {
        case "M":
          return mapper.readValue(bytes, ChatEvent.ChatMessage.class);
        case "J":
          return mapper.readValue(bytes, ChatEvent.JoinRoom.class);
        case "L":
          return mapper.readValue(bytes, ChatEvent.LeaveRoom.class);
        default:
          throw new RuntimeException("Unknown manifest: " + manifest);
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }
}

File: samples/scala/clustered-chat/app/chat/ChatEventSerializer.scala
package chat

import org.apache.pekko.actor.ExtendedActorSystem
import org.apache.pekko.serialization.BaseSerializer
import org.apache.pekko.serialization.SerializerWithStringManifest
import play.api.libs.json.Json

/**
 * Since messages sent through distributed pubsub go over Pekko remoting, they need to be
 * serialized. This serializer serializes them as JSON.
 */
class ChatEventSerializer(val system: ExtendedActorSystem) extends SerializerWithStringManifest with BaseSerializer {
  override def manifest(o: AnyRef): String = o match {
    case _: ChatMessage => "M"
    case _: JoinRoom    => "J"
    case _: LeaveRoom   => "L"
    case other          => sys.error("Don't know how to serialize " + other)
  }

  override def toBinary(o: AnyRef): Array[Byte] = {
    val json = o match {
      case cm: ChatMessage => Json.toJson(cm)
      case jr: JoinRoom    => Json.toJson(jr)
      case lr: LeaveRoom   => Json.toJson(lr)
      case other           => sys.error("Don't know how to serialize " + other)
    }
    Json.toBytes(json)
  }

  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = {
    val json = Json.parse(bytes)
    manifest match {
      case "M"   => json.as[ChatMessage]
      case "J"   => json.as[JoinRoom]
      case "L"   => json.as[LeaveRoom]
      case other => sys.error("Unknown manifest " + other)
    }
  }
}
